Introduce STPCardParams

jflinter opened the pull request:
This splits `STPCard`, which was previously used for both creating requests to the Stripe API and serializing responses from it, into `STPCardParams` and `STPCard`, respectively. (Same for `STPBankAccount`.) For compatibility reasons, `STPCard` inherits from `STPCardParams` - this way existing code won't break, although it will trigger deprecation warnings.
r? @benzguo

benzguo commented on Stripe/PublicHeaders/STPBankAccount.h:
```diff
@@ -7,19 +7,17 @@
 
 /**
- *  The account number for the bank account. Currently must be a checking account.
+ *  The last 4 digits of the bank account's account number, if it's been set, otherwise nil.
```
"otherwise nil" contradicts the `nonnull`

benzguo commented on Stripe/PublicHeaders/STPBankAccountParams.h:
```diff
@@ -0,0 +1,42 @@
+/**
+ *  The last 4 digits of the bank account's account number, if it's been set, otherwise nil.
+ */
+- (nullable NSString *)last4;
```
do we need the `last4` getter, if this object just gets sent to the API?

benzguo replied:
(still not sure why we need to expose this in the params object)

benzguo commented on Stripe/PublicHeaders/STPBankAccountParams.h:
```diff
@@ -0,0 +1,42 @@
+@property (nonatomic, copy, nullable) NSString *routingNumber;
+
+/**
+ *  The country the bank account is in.
```
maybe specify that this is a country code / ISO-X ?

benzguo commented on Stripe/PublicHeaders/STPBankAccountParams.h:
```diff
@@ -0,0 +1,42 @@
+/**
+ *  The default currency for the bank account.
+ */
+@property (nonatomic, copy, nullable) NSString *currency;
```
maybe specify that this is a currency code / ISO-X ?

benzguo commented on Stripe/PublicHeaders/STPCardParams.h:
```diff
@@ -0,0 +1,90 @@
+@property (nonatomic) NSUInteger expYear;
+
+/**
+ *  The card's security code, found on the back. This will be nil for cards retrieved from the Stripe API.
```
remove "This will be nil for cards retrieved from the Stripe API." ?

benzguo commented on Stripe/PublicHeaders/STPCardParams.h:
```diff
@@ -0,0 +1,90 @@
+@property (nonatomic, copy, nullable) NSString *addressCountry;
+
+/**
+ *  This is only applicable when tokenizing debit cards to issue payouts to managed accounts. You should not set it otherwise. The card can then be used as a transfer destination for funds in this currency.
```
specify that this is a currency code?

benzguo commented on Stripe/PublicHeaders/UI/STPPaymentCardTextField.h:
```diff
@@ -157,7 +157,7 @@
 /**
  *  Convenience method to create a STPCard from the currently entered information. Will return nil if not valid.
```
nit: kinda confusing docstring, this isn't a method

jflinter commented:
Mind re-reviewing, @benzguo?
I added a lot more code to this PR, but the card -> cardparams stuff is unchanged. I added two new protocols, `STPFormEncodable` and `STPAPIResponseDecodable`, which basically are the ~same as `StripeAPITranslatable`.

benzguo commented on Stripe/PublicHeaders/STPAPIResponseDecodable.h:
```diff
@@ -0,0 +1,15 @@
+
+@protocol STPAPIResponseDecodable <NSObject>
+
++ (nullable instancetype)decodedObjectFromAPIResponse:(nonnull NSDictionary *)response;
```
nice.

benzguo commented on Stripe/PublicHeaders/STPBankAccountParams.h:
```diff
@@ -0,0 +1,43 @@
+@property (nonatomic, copy, nullable) NSString *country;
+
+/**
+ *  The default currency for the bank account.
```
nit: "Three-letter ISO currency code representing the default currency for the bank account."

benzguo commented on Stripe/PublicHeaders/STPFormEncodable.h:
```diff
@@ -0,0 +1,16 @@
+@protocol STPFormEncodable <NSObject>
+
+- (NSString *)rootObjectName;
+- (NSDictionary *)propertyNamesToFormFieldNamesMapping;
```
(no need to change) 
I've taken to just calling dictionaries `keyToValue`, e.g. `propertyNameToFormFieldName`

benzguo commented:
looks good! 
my only question may just be me missing something – why are we exposing `last4` in the params objects?

jflinter commented:
Backwards compatibility - that code existed with that behavior on
STPCard/STPBankAccount, and I don't want to break things (even if the
behavior is a little silly). The true breaking change will be to make
STPCard not inherit from STPCardParams, at which point we'll remove this.

On Wed, Oct 14, 2015 at 4:07 PM, Ben Guo <notifications@github.com> wrote:

> looks good!
> my only question may just be me missing something – why are we exposing
> last4 in the params objects?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/stripe/stripe-ios/pull/247#issuecomment-148180617>.
>

benzguo commented:
ah, makes sense. :ship:
